Doc/c-api/iter.rst

```diff
@@ -19,40 +20,44 @@ There are two functions specifically for working with iterators.
 
    .. versionadded:: 3.10
 
-.. c:function:: PyObject* PyIter_Next(PyObject *o)
+.. c:function:: int PyIter_NextItem(PyObject *iter, PyObject **item)
 
-   Return the next value from the iterator *o*.  The object must be an iterator
-   according to :c:func:`PyIter_Check` (it is up to the caller to check this).
-   If there are no remaining values, returns ``NULL`` with no exception set.
-   If an error occurs while retrieving the item, returns ``NULL`` and passes
-   along the exception.
+   Return ``1`` and set *item* to a :term:`strong reference` of the
+   next value of the iterator *iter* on success.
+   Return ``0`` and set *item* to ``NULL`` if there are no remaining values.
+   Return ``-1``, set *item* to ``NULL`` and set an exception on error.
 
-To write a loop which iterates over an iterator, the C code should look
-something like this::
+   Iterate over *iter* using the following pattern::
 
-   PyObject *iterator = PyObject_GetIter(obj);
-   PyObject *item;
+      PyObject *iter = PyObject_GetIter(obj);
+      if (iter == NULL) {
+          goto error;
+      }
 
-   if (iterator == NULL) {
-       /* propagate error */
-   }
+      PyObject *item = NULL;
+      while (PyIter_NextItem(iter, &item)) {
+          if (item == NULL) {
+              Py_DECREF(iter);
+              goto error;
+          }
+          do_something(item);
+          Py_DECREF(item);
+      }
+      Py_DECREF(iter);
 
-   while ((item = PyIter_Next(iterator))) {
-       /* do something with item */
-       ...
-       /* release reference when done */
-       Py_DECREF(item);
-   }
+   .. versionadded:: 3.14
 
-   Py_DECREF(iterator);
+.. c:function:: PyObject* PyIter_Next(PyObject *o)
 
-   if (PyErr_Occurred()) {
-       /* propagate error */
-   }
-   else {
-       /* continue doing useful work */
-   }
+   This is an older version of :c:func:`!PyIter_NextItem`,
```

Comment:
Why not adding a link to PyIter_NextItem?

Reply:
The link is on the following line. We normally do not link twice to the same anchor in a paragraph. I found the second mention of `PyIter_NextItem` to be a better place for the link than the first mention.